pkg/sql/sqlstats/sslocal/sslocal_test: TestSQLStatsLatencyInfo failed

pkg/sql/sqlstats/sslocal/sslocal_test.TestSQLStatsLatencyInfo [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9706826?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9706826?buildTab=artifacts#/) on master @ [7e4e0b195cd61da6cd7a719a5b9aa2e84f68d475](https://github.com/cockroachdb/cockroach/commits/7e4e0b195cd61da6cd7a719a5b9aa2e84f68d475):


```
=== RUN   TestSQLStatsLatencyInfo
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/679f97a0c7f1bc1cdbf4b259e3dc42c0/logTestSQLStatsLatencyInfo717390305
    test_log_scope.go:79: use -show-logs to present logs inline
=== CONT  TestSQLStatsLatencyInfo
    sql_stats_test.go:1525: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/679f97a0c7f1bc1cdbf4b259e3dc42c0/logTestSQLStatsLatencyInfo717390305
--- FAIL: TestSQLStatsLatencyInfo (7.90s)
=== RUN   TestSQLStatsLatencyInfo/select_sleep
    sql_stats_test.go:1519: 
        	Error Trace:	pkg/sql/sqlstats/sslocal/sslocal_test_test/pkg/sql/sqlstats/sslocal/sql_stats_test.go:1519
        	Error:      	"0.212056587" is not less than or equal to "0.2"
        	Test:       	TestSQLStatsLatencyInfo/select_sleep
    --- FAIL: TestSQLStatsLatencyInfo/select_sleep (0.10s)
=== RUN   TestSQLStatsLatencyInfo/select_sleep#01
    sql_stats_test.go:1519: 
        	Error Trace:	pkg/sql/sqlstats/sslocal/sslocal_test_test/pkg/sql/sqlstats/sslocal/sql_stats_test.go:1519
        	Error:      	"0.212056587" is not less than or equal to "0.2"
        	Test:       	TestSQLStatsLatencyInfo/select_sleep#01
    --- FAIL: TestSQLStatsLatencyInfo/select_sleep#01 (0.14s)
=== RUN   TestSQLStatsLatencyInfo/select_sleep#02
    sql_stats_test.go:1519: 
        	Error Trace:	pkg/sql/sqlstats/sslocal/sslocal_test_test/pkg/sql/sqlstats/sslocal/sql_stats_test.go:1519
        	Error:      	"0.212056587" is not less than or equal to "0.2"
        	Test:       	TestSQLStatsLatencyInfo/select_sleep#02
    --- FAIL: TestSQLStatsLatencyInfo/select_sleep#02 (0.11s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cluster-observability
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSQLStatsLatencyInfo.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-27187

--- a/pkg/sql/sqlstats/sslocal/BUILD.bazel
+++ b/pkg/sql/sqlstats/sslocal/BUILD.bazel
@@ -73,6 +73,7 @@ go_test(
         "//pkg/util/leaktest",
         "//pkg/util/log",
         "//pkg/util/mon",
+        "//pkg/util/timeutil",
         "@com_github_jackc_pgx_v4//:pgx",
         "@com_github_lib_pq//:pq",
         "@com_github_stretchr_testify//require",

--- a/pkg/sql/sqlstats/sslocal/sql_stats_test.go
+++ b/pkg/sql/sqlstats/sslocal/sql_stats_test.go
@@ -47,6 +47,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/jackc/pgx/v4"
 	"github.com/lib/pq"
 	"github.com/stretchr/testify/require"
@@ -1467,31 +1468,26 @@ func TestSQLStatsLatencyInfo(t *testing.T) {
 		name        string
 		statement   string
 		fingerprint string
-		latencyMax  float64
 	}{
 		{
 			name:        "select on table",
 			statement:   "SELECT * FROM t1",
 			fingerprint: "SELECT * FROM t1",
-			latencyMax:  1,
 		},
 		{
-			name:        "select sleep",
+			name:        "select sleep(0.06)",
 			statement:   "SELECT pg_sleep(0.06)",
 			fingerprint: "SELECT pg_sleep(_)",
-			latencyMax:  0.2,
 		},
 		{
-			name:        "select sleep",
+			name:        "select sleep(0.1)",
 			statement:   "SELECT pg_sleep(0.1)",
 			fingerprint: "SELECT pg_sleep(_)",
-			latencyMax:  0.2,
 		},
 		{
-			name:        "select sleep",
+			name:        "select sleep(0.07)",
 			statement:   "SELECT pg_sleep(0.07)",
 			fingerprint: "SELECT pg_sleep(_)",
-			latencyMax:  0.2,
 		},
 	}
 
@@ -1500,9 +1496,13 @@ func TestSQLStatsLatencyInfo(t *testing.T) {
 	var p50 float64
 	var p90 float64
 	var p99 float64
+	stopwatch := timeutil.NewStopWatch()
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
+
+			stopwatch.Start()
 			testConn.Exec(t, tc.statement)
+			stopwatch.Stop()
 
 			rows := testConn.QueryRow(t, "SELECT statistics -> 'statistics' -> 'latencyInfo' ->> 'min',"+
 				"statistics -> 'statistics' -> 'latencyInfo' ->> 'max',"+
@@ -1513,13 +1513,14 @@ func TestSQLStatsLatencyInfo(t *testing.T) {
 				"AND metadata ->> 'query'=$2", appName, "SELECT * FROM t1")
 			rows.Scan(&min, &max, &p50, &p90, &p99)
 
-			require.Positive(t, min)
-			require.Positive(t, max)
-			require.GreaterOrEqual(t, max, min)
-			require.LessOrEqual(t, max, tc.latencyMax)
-			require.GreaterOrEqual(t, p99, p90)
-			require.GreaterOrEqual(t, p90, p50)
-			require.LessOrEqual(t, p99, max)
+			require.Positivef(t, min, "expected min latency %f greater than 0", min)
+			require.Positivef(t, max, "expected max latency %f greater than 0", max)
+			require.GreaterOrEqualf(t, max, min, "expected max latency %f greater than min latency %f", max, min)
+			require.LessOrEqualf(t, max, stopwatch.Elapsed().Seconds(), "expected max latency %f less or equal %f",
+				max, stopwatch.Elapsed().Seconds())
+			require.GreaterOrEqualf(t, p99, p90, "expected p99 latency %f greater or equal p90 latency %f", p99, p90)
+			require.GreaterOrEqualf(t, p90, p50, "expected p90 latency %f greater or equal p50 latency %f", p90, p50)
+			require.LessOrEqualf(t, p99, max, "expected p99 latency %f less or equal max latency %f", p99, max)
 		})
 	}
 }

